### PR TITLE
[FLINK-19219] Run JobManager initialization in separate thread to make it cancellable

### PR DESCRIPTION
**Pull request by rmetzger**
## What is the purpose of the change

If a job gets stuck in the INITIALIZING state, we can not properly cancel it in this stage.
What effectively happens is that we are waiting for the initialization to be over before cancelling the job.

This change runs the initialization in a separate thread, that we can interrupt in case the user issues a cancel call.

This approach does not guarantee that the initialization is cooperatively aborting the current operation. If it is not cooperating, we will wait and follow the old approach of cancelling.

## Brief change log

- Introduce `InitializingJobManagerRunner` class that hides the thread handling
- adjusted `Dispatcher` and `DispatcherJob` to cooperate with the new class 

## Verifying this change

I adjusted some existing tests. 
This is a draft PR, once it seems to be ok, I'll add a test for `InitializingJobManagerRunner`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / *no*)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / *no*)
  - The serializers: (yes / *no* / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / *no* / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (*no* / no / don't know)
  - The S3 file system connector: (yes / *no* / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / *no*)
  - If yes, how is the feature documented? (not applicable / docs / *JavaDocs* / not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java [hunk 1]
```diff
@@ -119,11 +119,17 @@ private DispatcherJob(
 						}
 					});
 				} else { // failure during initialization
-					final Throwable strippedThrowable = ExceptionUtils.stripCompletionException(throwable);
+					JobStatus finalStatus = JobStatus.FAILED;
+					Throwable strippedThrowable = ExceptionUtils.stripCompletionException(throwable);
+					if (throwable instanceof JobInitializationCancelledException) {
```
[the remark this answers is not in the document]
Comment (rmetzger) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java:124]: but there is no jobManagerRunner available if it's initialization failed exceptionally.

I will add a unit test that ensures the correct exception types to avoid accidental changes.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java [hunk 2]
```diff
@@ -165,17 +171,10 @@ private DispatcherJob(
 			if (isInitialized()) {
 				return getJobMasterGateway().thenCompose(jobMasterGateway -> jobMasterGateway.cancel(timeout));
 			} else {
-				log.info("Cancellation during initialization requested for job {}. Job will be cancelled once JobManager has been initialized.", jobId);
+				log.info("Cancellation during initialization for job {}.", jobId);
```
[the remark this answers is not in the document]
Comment (rmetzger) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java:174]: Since it is quite rare to cancel jobs during initialization, I don't believe this log message will hurt

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/InitializingJobManagerRunner.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	private final Thread initializationRunner;
+
+	private final Object lock = new Object();
+
+	@GuardedBy("lock")
+	private CompletableFuture<Acknowledge> cancellationFuture;
+
+	public InitializingJobManagerRunner(SupplierWithException<JobManagerRunner, JobInitializationException> initializationRunnable, JobID jobID) {
+		String threadName = "jobmanager-initialization-" + jobID;
+		this.initializationRunner = new Thread(() -> {
```
[the remark this answers is not in the document]
Comment (rmetzger) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/InitializingJobManagerRunner.java:53]: I thought I can only set the interrupt flag of a thread if I control it myself. Is there a way of setting the flag when using a pool?

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/InitializingJobManagerRunner.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+					jobManagerRunnerFuture.completeExceptionally(new JobInitializationCancelledException("Job initialization has been cancelled", throwable));
+					cancellationFuture.complete(Acknowledge.get());
+				}
+			}
+		}
+	}
+
+	public CompletableFuture<Acknowledge> cancelInitialization() {
+		synchronized (lock) {
+			initializationRunner.interrupt();
```
[the remark this answers is not in the document]
Comment (rmetzger) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/InitializingJobManagerRunner.java:96]: The thread will also stop if it enters a (properly implemented) blocking operation.

The classical case we are targeting is that a remote system (triggered by initializeOnMaster(), or some state restore operation) is not available. The code will be in blocking operations all the time (waiting for network, sleeping, ...).
The bigger question is whether the thread is cooperative in reacting to the interrupt. Often, interrupt exceptions are ignored.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java:124]: I meant `this.initializingJobManagerRunner`.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java:174]: That's not really an answer to my question, and imo further reinforces that it shouldn't be logged on info.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/InitializingJobManagerRunner.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/InitializingJobManagerRunner.java:53]: hmm...`ExecutorService#submit` returns a `Future` on which you can call `cancel`. `cancel` also has a parameter to interrupt the thread.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/InitializingJobManagerRunner.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/InitializingJobManagerRunner.java:96]: Since we intend to move the `initializeOnMaster` execution to the IO executor in FLINK-19069, what's your plan to handle this? (do we even need FLINK-19069 after this PR?)

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (rmetzger) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java:124]: I see. I will look into it!

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (rmetzger) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/DispatcherJob.java:174]: Let me try again: I believe it is relevant enough because we are interrupting the initialization thread, potentially leading to all kinds of weird errors. I should change the log message a bit to make that more explicit.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/InitializingJobManagerRunner.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (rmetzger) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/InitializingJobManagerRunner.java:53]: You are right. I assumed Futures are never really cancellable, because `CompletableFuture.cancel()` has the following javadoc:
```
     * @param mayInterruptIfRunning this value has no effect in this
     * implementation because interrupts are not used to control
     * processing.
```

I will use the ioExecutor of the Dispatcher.

File: flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/InitializingJobManagerRunner.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (rmetzger) [reply flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/InitializingJobManagerRunner.java:96]: FLINK-19069 is about finalizeOnMaster, not initializeOnMaster.

**Comment by rmetzger** [comment]
I'm closing this PR for now. I suggest to re-open it if users ask for an improved cancellation behavior.
